src/Workspaces/Core/Portable/FindSymbols/FindLiterals/FindLiteralsSearchEngine.cs

```diff
@@ -0,0 +1,201 @@
+                }
+            }
+            finally
+            {
+                await _progressTracker.ItemCompletedAsync().ConfigureAwait(false);
+            }
+        }
+
+        private async Task FindReferencesWorkerAsync()
+        {
+            var count = _solution.Projects.SelectMany(p => p.DocumentIds).Count();
+            await _progressTracker.AddItemsAsync(count).ConfigureAwait(false);
+
+            foreach (var project in _solution.Projects)
+            {
+                _cancellationToken.ThrowIfCancellationRequested();
+
+                var documentTasks = new List<Task>();
+                foreach (var document in project.Documents)
+                {
+                    documentTasks.Add(ProcessDocumentAsync(document));
+                }
+
+                await Task.WhenAll(documentTasks).ConfigureAwait(false);
+            }
+        }
+
+        private async Task ProcessDocumentAsync(Document document)
+        {
+            try
+            {
+                await ProcessDocumentWorkerAsync(document).ConfigureAwait(false);
+            }
+            finally
+            {
+                await _progressTracker.ItemCompletedAsync().ConfigureAwait(false);
+            }
+        }
+
+        private async Task ProcessDocumentWorkerAsync(Document document)
+        {
+            var index = await SyntaxTreeIndex.GetIndexAsync(
+                document, _cancellationToken).ConfigureAwait(false);
+
+            if (_searchKind == SearchKind.StringLiterals)
+            {
+                if (index.ProbablyContainsStringValue(_stringValue))
+                {
+                    await SearchDocumentAsync(document).ConfigureAwait(false);
+                }
+            }
+            else if (index.ProbablyContainsInt64Value(_longValue))
+            {
+                await SearchDocumentAsync(document).ConfigureAwait(false);
+            }
+        }
+
+        private async Task SearchDocumentAsync(Document document)
+        {
+            _cancellationToken.ThrowIfCancellationRequested();
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var root = await document.GetSyntaxRootAsync(_cancellationToken).ConfigureAwait(false);
+
+            var matches = ArrayBuilder<SyntaxToken>.GetInstance();
+            ProcessNode(syntaxFacts, root, matches);
+
+            foreach (var token in matches)
+            {
+                await _progress.OnReferenceFoundAsync(document, token.Span).ConfigureAwait(false);
+            }
+        }
+
+        private void ProcessNode(
+            ISyntaxFactsService syntaxFacts, SyntaxNode node, ArrayBuilder<SyntaxToken> matches)
+        {
+            _cancellationToken.ThrowIfCancellationRequested();
+            foreach (var child in node.ChildNodesAndTokens())
+            {
+                if (child.IsNode)
+                {
+                    ProcessNode(syntaxFacts, child.AsNode(), matches);
```

Comment:
I don't think this needs to be recursive - can't you just use [`DescendantTokens()`](http://source.roslyn.io/Microsoft.CodeAnalysis/R/37f36c9babc896b1.html)?